### PR TITLE
Add users stats http api to ingester

### LINKED ISSUE #6144: Alternative to scaling down ingesters
**Is your feature request related to a problem? Please describe.**
Today, scaling down ingesters is a complicated and highly manual process. As described on the [website](https://cortexmetrics.io/docs/guides/ingesters-scaling-up-and-down/#scaling-down), scaling down requires ensuring that blocks are flushed to storage and that queries use the stored data set to 0s (.query-store-after). However, this approach is not suitable for all use cases, as in some scenarios, we want to utilize ingesters for querying as well, to improve request performance.

**Describe the solution you'd like**
Automating the scale down of ingesters is not a trivial task. It is desirable for ingesters to have a mechanism that allows users to scale them down gradually without missing data.

A proposed solution is to introduce a new state for ingesters called READONLY. In this state, ingesters cannot receive data, meaning all Push requests would fail, but they can still accept query data. Cortex would use the ring operation to filter out the correct ingesters by state, allowing the distributor and query/ruler to use the appropriate set of ingesters.
```
Write = NewOp([]InstanceState{ACTIVE})
Read = NewOp([]InstanceState{ACTIVE, PENDING, LEAVING, JOINING, READONLY})
```

To enable users to set an ingester to READONLY mode, ingesters would have a new API that allows them to transition to READONLY or ACTIVE. It would be permissible for an ingester to return to ACTIVE mode as a way to cancel a scale down if needed.
```
a.indexPage.AddLink(SectionDangerous, "/ingester/mode", "Change Ingester mode on ring")
```

Furthermore, to allow ingesters to be safely removed from the ring, they would also have a new API that lets users know which blocks an ingester has loaded. The idea is that when an ingester has deleted all blocks, it can be stopped.
```
a.indexPage.AddLink(SectionDangerous, "/ingester/blocks", "List blocks on ingesters")
```

This approach introduces a new READONLY state for ingesters, enabling a controlled scale down process without data loss. Users can transition ingesters to READONLY mode, preventing new data ingestion while allowing queries on existing data. Once an ingester has deleted all its blocks, it can be safely stopped and removed from the ring.

**Describe alternatives you've considered**
1. Using the LEAVING state as READONLY. This was discarded as the LEAVING state already has multiple logics and premises on why the pod is in that state, which could make the code more confusing.

2. Not having the /ingester/blocks endpoint and using the .query-store-after configuration to scale down ingesters. While this can still work, it adds complexity for the user as they would need to track the time, ensure the configuration hasn't changed, and account for failures in ingesters pushing blocks to storage.

**Additional context**

### PR DESCRIPTION
**Pull request by danielblando**
**What this PR does**:
As mentioned in the feature, it would be nice if we know ingesters are empty to safely scale it down.
This extends the existent api on distributors `/distributors/all_user_stats` to also ingesters `/ingester/all_user_stats`.
It was also added to the api the info about `Loaded blocks` which can be used to scale down ingesters.

Example of the API on ingester
Screenshot 2024-08-27 at 11 47 24 PM

**Which issue(s) this PR fixes**:
Fixes #6144

**Checklist**
- [X] Tests updated
- [ ] Documentation added
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by alanprot — APPROVED** [review]

File: pkg/ingester/http_admin.go [hunk 1]
```diff
@@ -53,31 +57,41 @@ const tpl = `
...
+}
+
+// UserStats models ingestion statistics for one user.
+type UserStats struct {
+	IngestionRate     float64 `json:"ingestionRate"`
+	NumSeries         uint64  `json:"numSeries"`
+	APIIngestionRate  float64 `json:"APIIngestionRate"`
+	RuleIngestionRate float64 `json:"RuleIngestionRate"`
+	ActiveSeries      uint64  `json:"activeSeries"`
+	LoadBlocks        uint64  `json:"loadBlocks"`
```
Comment (yeya24) [inline pkg/ingester/http_admin.go:89]: This is unclear to me. I see it just uses `db.Blocks()`. Does it include head block as well?
How we gonna use this data to help with scale down. Do we check `LoadBlocks==0`?

File: pkg/api/api.go [hunk 2]
```diff
@@ -284,19 +284,23 @@ type Ingester interface {
...
 
+	a.indexPage.AddLink(SectionAdminEndpoints, "/ingester/all_user_stats", "Usage Statistics")
+
 	a.indexPage.AddLink(SectionDangerous, "/ingester/flush", "Trigger a Flush of data from Ingester to storage")
 	a.indexPage.AddLink(SectionDangerous, "/ingester/shutdown", "Trigger Ingester Shutdown (Dangerous)")
 	a.indexPage.AddLink(SectionDangerous, "/ingester/renewTokens", "Renew Ingester Tokens (10%)")
 	a.RegisterRoute("/ingester/flush", http.HandlerFunc(i.FlushHandler), false, "GET", "POST")
 	a.RegisterRoute("/ingester/shutdown", http.HandlerFunc(i.ShutdownHandler), false, "GET", "POST")
 	a.RegisterRoute("/ingester/renewTokens", http.HandlerFunc(i.RenewTokenHandler), false, "GET", "POST")
+	a.RegisterRoute("/ingester/all_user_stats", http.HandlerFunc(i.AllUserStatsHandler), false, "GET")
```
Comment (yeya24) [inline pkg/api/api.go:303]: We need to document the new API

File: CHANGELOG.md [hunk 3]
```diff
@@ -3,6 +3,7 @@
 ## master / unreleased
 
 * [ENHANCEMENT] Ruler: Add new ruler metric `cortex_ruler_rule_groups_in_store` that is the total rule groups per tenant in store, which can be used to compare with `cortex_prometheus_rule_group_rules` to count the number of rule groups that are not loaded by a ruler. #5869
+* [ENHANCEMENT] Ingester: Add new API `/ingester/all_user_stats` #6178
```
Comment (yeya24) [inline CHANGELOG.md:6]: Can we add more info about the API? At least describing what it does

File: pkg/ingester/http_admin.go (hunk 1 above)
Comment (alanprot) [reply pkg/ingester/http_admin.go:89]: This should be "loadedBlocks" ?

**Comment by danielblando** [comment]
Added api docs

Screenshot 2024-09-02 at 6 06 56 PM

**Review by yeya24 — APPROVED** [review]
Thanks

**Pull request merged**